src/libcephfs.cc

```diff
@@ -37,11 +37,34 @@
 #define DEFAULT_UMASK 002
 
 static mode_t umask_cb(void *);
-ceph::async::io_context_pool icp;
+namespace {
+// Set things up this way so we don't start up threads until mount and
+// kill them off when the last mount goes away, but are tolerant to
+// multiple mounts of overlapping duration.
+std::shared_ptr<ceph::async::io_context_pool> get_icp(CephContext* cct)
+{
+  static std::weak_ptr<ceph::async::io_context_pool> icwp;
+
+  auto icp = icwp.lock();
```

Comment:
My understanding is that this is not thread-safe. The `weak_ptr` either needs to be protected by a lock or access needs to be atomic. We've done something like this elsewhere in Ceph: https://github.com/ceph/ceph/blob/6c805c640527f9600fe04369ac4a5d8ec8583707/src/common/ceph_context.h#L272-L279

This is available in C++20 but that doesn't help us now: https://en.cppreference.com/w/cpp/memory/weak_ptr/atomic2

Reply:
Fixed